crates/lib3h/src/rrdht_util/r_value.rs

```diff
@@ -0,0 +1,335 @@
+        self
+    }
+
+    /// push a new PeerRecord into this record set
+    pub fn push_peer_record(mut self, peer_record: RValuePeerRecord) -> Self {
+        if !self
+            .arc_of_included_peer_records
+            .contains_location(peer_record.storage_arc.center())
+        {
+            panic!("peer record does not fit within arc_of_included_peer_records");
+        }
+        self.peer_records.push(peer_record);
+        self
+    }
+}
+
+/// Interpolate qualified redundancy in the network given our current
+/// knowledge of peers' claimed storage arcs.
+/// For a given storage arc (should match *this* agent's storage arc),
+/// if all the peers we know about are actually storing what they claim to be
+/// we can interpolate what the whole network "R" value is
+/// also taking into account our experience of those agents' reachability.
+pub fn interpolate_r_value_for_given_arc(peer_record_set: &RValuePeerRecordSet) -> f64 {
+    let mut running_avg_coverage = 0.0_f64;
+
+    for (i, record) in peer_record_set.peer_records.iter().enumerate() {
+        // what is the total coverage factor for this record?
+        let coverage =
+            (record.storage_arc.length() as f64 / ARC_LENGTH_MAX as f64) * record.uptime_0_to_1;
+
+        // mix it into our running average
+        running_avg_coverage = (running_avg_coverage * i as f64 + coverage) / (i as f64 + 1.0);
+    }
+
+    // given our known arc, estimate the total node count
+    let est_node_count = peer_record_set.peer_records.len() as f64
+        * (1.0
+            / (peer_record_set.arc_of_included_peer_records.length() as f64
+                / ARC_LENGTH_MAX as f64));
+
+    // get the estimated total coverage a.k.a r-value
+    running_avg_coverage * est_node_count
+}
+
+/// As an agent, we need to set our storage arc radius to something
+/// on the one hand, if we have enough storage and network bandwidth,
+/// we might want to set it as high as possible. On the other hand, we
+/// might want to conserve resources / power. If our network is healthy,
+/// we can scale back a bit to conserve, if it is in danger, we may want
+/// to scale up for safety. If it is anywhere in the middle, we probably
+/// want to maintain stability so we're not thrashing our resources.
+///
+/// Given current network conditions (within the sample slice given)
+/// this function will return a recommended storage arc radius.
+/// RADIUS not length. Please use `Arc::new_radius(0.into(), radius).length()`
+/// if you need to determine the length for a given radius..
+///
+/// If you provide a current_arc_radius, and the network is anywhere near
+/// stable, this function will probably return your current radius back.
+/// Otherwise it will pick a maintenance target right in the middle.
+///
+/// If the network is over replicated, this function may pick a small radius.
+/// If the network is unhealthy, or immature this function may pick a large
```

Comment:
Suggest "under replicated" instead of "unhealthy" here. The phrase "network is unhealthy" could be misinterpreted many ways (e.g. too many network partitions, malicious actors, high latency).